From: =?utf-8?q?Ond=C5=99ej_Jirman?= <megi@xff.cz>
Date: Wed, 7 Oct 2020 02:11:35 +0200
Subject: [PATCH] clk: sunxi-ng: sun50i-a64: Switch parent of MIPI-DSI to
 periph0(1x)

This makes video0(1x) clock less constrained, and improves compatibility
with external monitors on Pinephone when using both internal display
and HDMI output at once.

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 drivers/clk/sunxi-ng/ccu-sun50i-a64.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/clk/sunxi-ng/ccu-sun50i-a64.c b/drivers/clk/sunxi-ng/ccu-sun50i-a64.c
index f03aaf4..191b33f 100644
--- a/drivers/clk/sunxi-ng/ccu-sun50i-a64.c
+++ b/drivers/clk/sunxi-ng/ccu-sun50i-a64.c
@@ -950,6 +950,8 @@ static struct ccu_rate_reset_nb sun50i_a64_pll_video0_reset_tcon0_nb = {
 	.common		= &pll_video0_clk.common,
 };
 
+#define CCU_MIPI_DSI_CLK 0x168
+
 static int sun50i_a64_ccu_probe(struct platform_device *pdev)
 {
 	void __iomem *reg;
@@ -967,6 +969,12 @@ static int sun50i_a64_ccu_probe(struct platform_device *pdev)
 
 	writel(0x515, reg + SUN50I_A64_PLL_MIPI_REG);
 
+	/* Set MIPI-DSI clock parent to periph0(1x), so that video0(1x) is free to change. */
+	val = readl(reg + CCU_MIPI_DSI_CLK);
+	val &= 0x30f;
+	val |= (2 << 8) | ((4 - 1) << 0); /* M-1 */
+	writel(val, reg + CCU_MIPI_DSI_CLK);
+
 	/* Force the parent of TCON0 to PLL-MIPI */
 	val = readl(reg + SUN50I_A64_TCON0_REG);
 	val &= ~GENMASK(26, 24);
